hermes/crates/cardano-chain-follower/src/metadata/cip509/mod.rs

```diff
@@ -0,0 +1,954 @@
+                    },
+                    Cip509Int::TxInputsHash => {
+                        cip509_metadatum.txn_inputs_hash =
+                            decode_bytes(d, "CIP509 txn inputs hash")?
+                                .try_into()
+                                .map_err(|_| {
+                                    decode::Error::message("Invalid data size of TxInputsHash")
+                                })?;
+                    },
+                    Cip509Int::PreviousTxId => {
+                        cip509_metadatum.prv_tx_id = Some(
+                            decode_bytes(d, "CIP509 previous tx ID")?
+                                .try_into()
+                                .map_err(|_| {
+                                    decode::Error::message("Invalid data size of PreviousTxId")
+                                })?,
+                        );
+                    },
+                    Cip509Int::ValidationSignature => {
+                        let validation_signature = decode_bytes(d, "CIP509 validation signature")?;
+                        if validation_signature.is_empty() || validation_signature.len() > 64 {
+                            return Err(decode::Error::message(
+                                "Invalid data size of ValidationSignature",
+                            ));
+                        }
+                        cip509_metadatum.validation_signature = validation_signature;
+                    },
+                }
+            } else {
+                // Handle the x509 chunks 10 11 12
+                let x509_chunks = X509Chunks::decode(d, ctx)?;
+                cip509_metadatum.x509_chunks = x509_chunks;
+            }
+        }
+        Ok(cip509_metadatum)
+    }
+}
+
+#[allow(clippy::module_name_repetitions)]
+impl Cip509 {
+    /// Decode and validate CIP509 metadata.
+    pub(crate) fn decode_and_validate(
+        decoded_metadata: &DecodedMetadata, _slot: u64, txn: &MultiEraTx,
+        raw_aux_data: &RawAuxData, _chain: Network, txn_idx: usize,
+    ) {
+        // Get the CIP509 metadata if possible
+        let Some(k509) = raw_aux_data.get_metadata(LABEL) else {
+            return;
+        };
+
+        let mut validation_report = ValidationReport::new();
+        let mut decoder = Decoder::new(k509.as_slice());
+
+        let mut cip509 = match Cip509::decode(&mut decoder, &mut ()) {
+            Ok(metadata) => metadata,
+            Err(e) => {
+                Cip509::default().validation_failure(
+                    &format!("Failed to decode CIP509 metadata: {e}"),
+                    &mut validation_report,
+                    decoded_metadata,
+                );
+                return;
+            },
+        };
+
+        // Validate transaction inputs hash
+        match cip509.validate_txn_inputs_hash(txn, &mut validation_report, decoded_metadata) {
```

Comment:
for better readability of this function, I would move such matches into the separate one

Reply:
I think it is better to do this way because it already calls another function that do the validation.